yjit_codegen.c

```diff
@@ -915,73 +917,170 @@ gen_opt_ge(jitstate_t* jit, ctx_t* ctx)
-    jne_ptr(cb, side_exit);
+        // if (SPECIAL_CONST_P(recv)) {
+        // Bail if receiver is not a heap object
+        test(cb, REG0, imm_opnd(RUBY_IMMEDIATE_MASK));
+        jnz_ptr(cb, side_exit);
+        cmp(cb, REG0, imm_opnd(Qfalse));
+        je_ptr(cb, side_exit);
+        cmp(cb, REG0, imm_opnd(Qnil));
+        je_ptr(cb, side_exit);
 
-    // Bail if idx is not a FIXNUM
-    mov(cb, REG1, idx_opnd);
-    test(cb, REG1, imm_opnd(RUBY_FIXNUM_FLAG));
-    jz_ptr(cb, side_exit);
+        // Bail if recv has a class other than ::Hash.
+        // BOP_AREF check above is only good for ::Hash.
+        mov(cb, REG1, mem_opnd(64, REG0, offsetof(struct RBasic, klass)));
+        mov(cb, REG0, const_ptr_opnd((void *)rb_cHash));
+        cmp(cb, REG0, REG1);
+        jit_chain_guard(JCC_JNE, jit, &starting_context, OPT_AREF_MAX_CHAIN_DEPTH, side_exit);
 
-    // Save YJIT registers
-    yjit_save_regs(cb);
+        // Call VALUE rb_hash_aref(VALUE hash, VALUE key).
+        {
+            // Write incremented pc to cfp->pc as the routine can raise and allocate
+            mov(cb, REG0, const_ptr_opnd(jit->pc + insn_len(BIN(opt_aref))));
+            mov(cb, member_opnd(REG_CFP, rb_control_frame_t, pc), REG0);
+
+            // About to change REG_SP which these operands depend on. Yikes.
+            mov(cb, R8, recv_opnd);
+            mov(cb, R9, idx_opnd);
+
+            // Write sp to cfp->sp since rb_hash_aref might need to call #hash on the key
+            if (ctx->sp_offset != 0) {
+                x86opnd_t stack_pointer = ctx_sp_opnd(ctx, 0);
+                lea(cb, REG_SP, stack_pointer);
+                mov(cb, member_opnd(REG_CFP, rb_control_frame_t, sp), REG_SP);
+                ctx->sp_offset = 0; // REG_SP now equals cfp->sp
+            }
 
-    mov(cb, RDI, recv_opnd);
-    sar(cb, REG1, imm_opnd(1)); // Convert fixnum to int
-    mov(cb, RSI, REG1);
-    call_ptr(cb, REG0, (void *)rb_ary_entry_internal);
+            yjit_save_regs(cb);
 
-    // Restore YJIT registers
-    yjit_load_regs(cb);
+            mov(cb, C_ARG_REGS[0], R8);
+            mov(cb, C_ARG_REGS[1], R9);
+            call_ptr(cb, REG0, (void *)rb_hash_aref);
 
-    x86opnd_t stack_ret = ctx_stack_push(ctx, T_NONE);
-    mov(cb, stack_ret, RAX);
+            yjit_load_regs(cb);
 
-    return YJIT_KEEP_COMPILING;
+            // Push the return value onto the stack
+            x86opnd_t stack_ret = ctx_stack_push(ctx, T_NONE);
+            mov(cb, stack_ret, RAX);
+        }
+
+        // Jump to next instruction. This allows guard chains to share the same successor.
+        {
+            ctx_t reset_depth = *ctx;
+            reset_depth.chain_depth = 0;
+
+            blockid_t jump_block = { jit->iseq, jit_next_insn_idx(jit) };
+
+            // Generate the jump instruction
+            gen_direct_jump(
+                &reset_depth,
+                jump_block
+            );
+        }
+        return YJIT_END_BLOCK;
```

Comment:
Note to self: consider factoring this out. There is now three sites.

Reply:
Yeah maybe we should have a `jump_next_insn` or something like that 🤔